model/canvas.cc

```diff
@@ -628,44 +628,40 @@ namespace minsky
 
   void Canvas::copyVars(const std::vector<VariablePtr>& v)
   {
-    auto group=model->addGroup(new Group);
-    float maxWidth=0, totalHeight=0;
+    float maxWidth=0, totalHeight=0, yCentre=0;
     vector<float> widths, heights;
-    for (auto i: v)
-      {
-        RenderVariable rv(*i);
-        widths.push_back(rv.width());
-        heights.push_back(rv.height());
-        maxWidth=max(maxWidth, widths.back());
-        totalHeight+=heights.back();
-      }
-    float y=group->y() - totalHeight;
-    for (size_t i=0; i<v.size(); ++i)
-      {
-        auto ni=v[i]->clone();
-        group->addItem(ni);
-        ni->rotation=0;
-        //ni->zoomFactor=group->zoomFactor;
-        ni->moveTo(group->x()+maxWidth - widths[i],
-                   y+heights[i]);
-        // variables need to refer to outer scope
-        if (ni->name()[0]!=':')
-          ni->name(':'+ni->name());
-        y+=2*heights[i];
-      }
-    // Stock and flow variables on Godley icons should not be copied as groups. For ticket 1039
-    selection.clear();    
-    auto copyOfItems=group->items;
-    for (auto& i: copyOfItems)
-      {		
-         model->addItem(i);			  
-         selection.ensureItemInserted(i);		 
-         assert(!i->ioVar());
-      } 
-    if (!copyOfItems.empty()) setItemFocus(copyOfItems[0]);       
-    group->clear();  
-    model->removeGroup(*group);
-    requestRedraw();        
+    // Throw error if no stock/flow vars on Godley icon. For ticket 1039 
+    if (!v.empty()) {    
+	  selection.clear();	
+      for (auto i: v)
+        {
+          RenderVariable rv(*i);
+          widths.push_back(rv.width());
+          heights.push_back(rv.height());
+          maxWidth=max(maxWidth, widths.back());
+          totalHeight+=heights.back();
+        }
+      float y=v[0]->y() - totalHeight;
+      for (size_t i=0; i<v.size(); ++i)
+        {
+          auto ni=v[i]->clone();
+          ni->rotation=0;
+          //ni->zoomFactor=group->zoomFactor;
+          ni->moveTo(v[0]->x()+maxWidth-v[i]->zoomFactor()*widths[i],
+                     y+heights[i]);
+          // variables need to refer to outer scope
+          if (ni->name()[0]!=':')
+            ni->name(':'+ni->name());
+          y+=2*v[i]->zoomFactor()*heights[i];
+          // Stock and flow variables on Godley icons should not be copied as groups. For ticket 1039
+          ItemPtr newI(ni);
+          selection.insertItem(model->addItem(newI));		 
+        }
+        // Item focus put on one of the copied vars that are still in selection. For ticket 1039
+        setItemFocus(model->findAny
```

Comment:
What about using selection.items[0]? Obviously check for empty selection first, and clear the itemFocus in that case.

Reply:
Thanks, done. I forgot about that member of selection.